cloudpickle/cloudpickle_fast.py

```diff
@@ -0,0 +1,541 @@
+        Pickler.__init__(self, file, protocol=protocol)
+        # map functions __globals__ attribute ids, to ensure that functions
+        # sharing the same global namespace at pickling time also share their
+        # global namespace at unpickling time.
+        self.globals_ref = {}
+
+        # Take into account potential custom reducers registered by external
+        # modules
+        self.dispatch_table = copyreg.dispatch_table.copy()
+
+        self.dispatch_table.update(self.dispatch)
+
+        # Pickling functions and classes cannot be customized using the
+        # dispatch_table: indeed, pickling an object using the dispatch_table
+        # works by invoking a reducer specific to the object's type. When the
+        # object is a class, its type is often ``type``, except when the class
+        # is an instance of another metaclass. In this cased, the metaclass
+        # will likely not be known in advance, and thus cannot be special-cased
+        # using an entry in the dispatch_table.
+
+        # The Pickler's reducer_override, among other things, allows us to
+        # register a reducer that will be called for any class, independently
+        # of its type.
+        self.proto = int(protocol)
+
+    def reducer_override(self, obj):
+        """Custom reducing callback for functions and classes
+
+        This function is the analog of a custom save_global. However, the C
+        Pickler API does not expose low-level instructions such as save or
+        write. Instead, we return a reduce value the Pickler will
+        internally serialize via save_reduce.
+        """
+        t = type(obj)
+        try:
+            is_anyclass = issubclass(t, type)
+        except TypeError:  # t is not a class (old Boost; see SF #502085)
+            is_anyclass = False
+
+        if is_anyclass:
+            return _class_reduce(obj)
+        elif isinstance(obj, types.FunctionType):
+            return self._function_reduce(obj)
+        else:
+            # fallback to save_global, including the Pickler's distpatch_table
+            return NotImplemented
+
+    # function reducers are defined as instance methods of CloudPickler
+    # objects, as they rely on a CloudPickler attribute (globals_ref)
+    def _dynamic_function_reduce(self, func):
+        """Reduce a function that is not pickleable via attribute lookup.
+        """
+        newargs = self._function_getnewargs(func)
+        state = _function_getstate(func)
+        return (types.FunctionType, newargs, state, None, None,
+                _function_setstate)
+
+    def _function_reduce(self, obj):
+        """Select the reducer depending on obj's dynamic nature
+
+        This functions starts by replicating save_global: trying to retrieve
+        obj from an attribute lookup of a file-backed module. If this check
+        fails, then a custom reducer is called.
+        """
+        # There no special handling for builtin pypy functions like in
```

Comment:
You are fine with that right (Follow up on https://github.com/cloudpipe/cloudpickle/pull/253#issuecomment-499615058)?

Reply:
If so actually I can clean up a little bit of unnecessary `pypy` handling codeblocks. Also, I wonder if pypy faces the limitation we have with the `CPython` pickle. Pickling seems pretty different.